src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs

```diff
@@ -31,42 +37,140 @@ public class AWSECSResourceDetector : IResourceDetector
+            resourceAttributes.AddRange(ExtractMetadataV4ResourceAttributes());
         }
         catch (Exception ex)
         {
             AWSXRayEventSource.Log.ResourceAttributesExtractException(nameof(AWSECSResourceDetector), ex);
         }
 
-        return resourceAttributes;
+        return new Resource(resourceAttributes);
+    }
+
+    internal static string GetStringOrThrow(JObject obj, string fieldName)
+    {
+        var value = obj[fieldName]?.ToString();
+        if (value == null)
+        {
+            throw new ArgumentNullException($"The expected '{fieldName}' field is missing");
+        }
+
+        return value;
     }
 
-    internal List<KeyValuePair<string, object>> ExtractResourceAttributes(string containerId)
+    internal static List<KeyValuePair<string, object>> ExtractResourceAttributes(string containerId)
     {
         var resourceAttributes = new List<KeyValuePair<string, object>>()
         {
-            new KeyValuePair<string, object>(AWSSemanticConventions.AttributeCloudProvider, "aws"),
-            new KeyValuePair<string, object>(AWSSemanticConventions.AttributeCloudPlatform, "aws_ecs"),
             new KeyValuePair<string, object>(AWSSemanticConventions.AttributeContainerID, containerId),
         };
 
         return resourceAttributes;
     }
 
+    internal static List<KeyValuePair<string, object>> ExtractMetadataV4ResourceAttributes()
+    {
+        var metadataV4Url = Environment.GetEnvironmentVariable(AWSECSMetadataURLV4Key);
+        if (metadataV4Url == null)
+        {
+            return new List<KeyValuePair<string, object>>();
+        }
+
+        var httpClientHandler = new HttpClientHandler();
+        var metadataV4ContainerResponse = ResourceDetectorUtils.SendOutRequest(metadataV4Url, "GET", null, httpClientHandler).Result;
+        var metadataV4TaskResponse = ResourceDetectorUtils.SendOutRequest($"{metadataV4Url.TrimEnd('/')}/task", "GET", null, httpClientHandler).Result;
+
+        var containerResponse = JObject.Parse(metadataV4ContainerResponse);
+        var taskResponse = JObject.Parse(metadataV4TaskResponse);
+
+        var containerArn = GetStringOrThrow(containerResponse, "ContainerARN");
+        var clusterArn = GetStringOrThrow(taskResponse, "Cluster");
+
+        if (!clusterArn.StartsWith("arn:"))
+        {
+            var baseArn = containerArn.Substring(containerArn.LastIndexOf(":"));
+            clusterArn = $"{baseArn}:cluster/{clusterArn}";
+        }
+
+        var launchType = GetStringOrThrow(taskResponse, "LaunchType") switch
+        {
+            string type when "ec2".Equals(type.ToLower()) => AWSSemanticConventions.ValueEcsLaunchTypeEc2,
+            string type when "fargate".Equals(type.ToLower()) => AWSSemanticConventions.ValueEcsLaunchTypeFargate,
+            _ => null,
+        };
+
+        if (launchType == null)
+        {
+            throw new ArgumentOutOfRangeException($"Unrecognized launch type '{taskResponse["LaunchType"]}'");
```

Comment:
I think that, from the PoV of semantic conventions, it would be exceptional:

```
aws.ecs.launchtype MUST be one of the following:

Value | Description
-- | --
ec2 | ec2
fargate | fargate
```

From https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/cloud_provider/aws/ecs/ .

Now, we might argue "But what if the Semantic Conventions are incomplete?" Today, I don't think they are. There is, technically, a _third_ launch type: `external`: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/launch_types.html. But, to the best of my knowledge, there is no Metadata endpoint v4 on external VMs in ECS, so it would not apply for this detector.

Reply:
Just in case I am wrong, I added some resilience in f35d9eb